README.md
SAGA: Skill to Action Generation for Agents
===========================================

Demo Quickstart
-------------
1. Make sure you have git, python-3.10/11, and poetry installed and OPENAI key setup (see below).
2. Clone this repo and change the directory to it.
3. run `poetry install` to install dependencies.
4. start a poetry shell with `poetry shell` to make sure you are using the correct python version and have the correct environment variables set.
5. run `python demos/space_colony/simulation.py`

B.Y.O.S. (Bring Your Own Simulation)
--------
SAGA is a library that can be used to generate actions for any simulation. [Our blog post shows
SAGA being used in our "Thistle Gulch" simulation](https://blog.fabledev.com/), which is full on 3D interactive simulation, but to
get started, you can use the included Space Colony demo as a starting point.

The Space Colony demo is a simple text-based simulation of a spaceship with a crew of 5 agents. It's included in this 
repo to make it easier to test it out and have an example to work from. See the section below for more details on
running it, as well as how it uses SAGA to generate actions.

Using as a Server (HTTP or SocketIO)
==================
SAGA can be used as a server that can be connected to via HTTP or SocketIO. This is useful if you want to interface
with SAGA from a simulation that is not written in Python (like the Thistle Gulch demo). To start the server, run:

`python -m fable_saga.server --type <http or socketio> --port <defaults to 8080> --host <defaults to localhost> --cors <defaults to *>`

Note that the server is for demo purposes and probably not secure, so you should only run it locally or on a secure network.
If you wanted to write your own server, see the `fable_saga.server` module for details on how that's being done.

HTTP
----

`python -m fable_saga.server --type http`

The HTTP server is a REST API that accepts POST requests to the `/` endpoint. The body of the request should be a JSON
object with the skills and context defined. You can optionally pass a "reference" that will be returned as well.

```
curl --request POST \
  --url http://localhost:8080/ \
  --header 'content-type: application/json' \
  --data '{
  "reference": "1234",
  "context": "You are a mouse",
  "skills": [{
    "name": "goto",
    "description": "go somewhere",
  	"parameters": {
      "location": "<str: where you want to go>"
    }
  }]
}'
```


The response looks like this:
```
{
  "actions": {
    "options": [
      {
        "skill": "goto",
        "parameters": {
          "location": "garden"
        }
      },
      ... all the other options
    ],
    "scores": [
      0.9,
      .. all the other scores
    ],
    "raw_prompt": <str: the raw prompt that was sent to the model, useful to debug.>,
    "raw_response": <str: the raw response from the model, useful if parsing failed and to debug.>,
    "llm_info": { ..this is specific to the model you are using, but openai models have this structure..
      "token_usage": { 
        "prompt_tokens": 164,
        "completion_tokens": 123,
        "total_tokens": 287
      },
      "model_name": "gpt-3.5-turbo"
    },
    "retries": 0, .. number of times the model was retried due to errors
    "error": null .. the error of the last try. null if no errors.
  },
  "error": null, .. the error either of the last try, or an error with the request itself.
  "reference": "1234" .. the reference you passed in
}
```

SocketIO
--------

`python -m fable_saga.server --type socketio`

The SocketIO server works similarly to the HTTP server, but it uses SocketIO instead.
You need to pass messages to the `generate-actions` event using the same JSON structure above, and the
response will be sent back to the same client that made the request in an async fashion.

The SocketIO server supports v4 of the SocketIO protocol, so you can use the SocketIO client of your choice.

When a client successfully connects to the server, you will see the following message:

`2023-12-22 12:50:07,972 - saga.server - INFO - connect:9CzowyUdrI1EOG0TAAAB`

Space Colony Demo
==================

The Space Colony demo is a simple text-based simulation of a spaceship with a crew of 5 agents.
There are various locations on the ship that the agents can move to and objects they can interact
with. All of this data we call "Meta-Memories" and they formatted and provided to the `fable_saga.Agent`
when requesting actions.

Generating Actions from Skills
-------------
Right now the agents and their spaceship is very simple, but with the few actions they have, they
can generate a lot of interesting behavior. The actions they can take are defined by "Skills" which
are listed in the `demos/space_colony/resources/skills.yaml` file. Here are two of them:

```Yaml
- name: go_to
  description: "Go to a location in the world"
  parameters:
    destination: "<str: persona_guid, item_guid, or location.name to go to>"
    goal: "<str: goal of the movement>"

- name: converse_with
  description: "Walk to another character and talk to them"
  parameters:
    persona_guid: "<str: guid of the persona to converse with. You cannot talk to yourself.>"
    topic: "<str: topic of the conversation>"
    context: "<str: lots of helpful details the conversation generator can use to
     generate a conversation. It only has access to the context and the topic
      you provide, so be very detailed.>"
    goal: "<str: goal of the conversation>"
```
It's important to note that SAGA itself doesn't know how to drive the simulation or characters. That
is left to the simulation itself. SAGA only knows how to generate actions based on the skills you
provide it. You can add more skills to the demo for instance, and SAGA will be able to generate
actions for them, but you will have to implement the logic for those actions in the demo simulation.
You can also use the `fable_saga.Agent` class outside the demo to generate actions for your own sim. 
This demo just makes it easy to see how the agents use these skills to generate actions.

Skills are used by SAGA to generate Action options. In the demo, they are returned to you interactively
on the command line, where you can see the options and choose one for the agent to take. The highest
scored action is always the first one in the list, so typing "0" and ENTER repeatedly will always choose the
highest scored action.

```
========== nyah_kobari ===========
  ROLE: As the head of security, Nyah ensures the safety of the cargo and crew from pirates,
     smugglers, and other potential threats. She also serves as a tactical advisor
     to Captain Sundeep and leads boarding and inspection procedures.
  LOCATION: Mess Hall
  ACTION: Idle
  MEMORIES:
  * 0m ago: Moved from crew_quarters_corridor to mess_hall with goal socialize with the crew
..
Generating actions for nyah_kobari using model gpt-3.5-turbo-1106...
...
#0 -- converse_with (0.8)
  persona_guid: elara_sundeep
  topic: ship's route planning
  context: Discussing potential alternate routes and navigation strategies
    to avoid asteroid fields and optimize fuel usage
  goal: gather input for route planning
#1 -- interact (0.7)
  item_guid: crew_photo_wall
  interaction: view_photos
  goal: reflecting on past memories and relationships with the crew
> Choose an item for nyah_kobari...
```

Memories
-------------
As the agents move around the ship and interact with objects and other agents, they generate memories
of those interactions. These memories are formatted as a list when passed in as context data to
the `fable_saga.Agent` class. The demo simulation keeps track of these memories, not SAGA.

The memories will continue to be generated as the agents interact with the world. This can add a lot of
tokens to the context data, so if you have any intention of running the demo for a long time, you will
want to implement some kind of memory management system to keep the context data from getting too large.

Large Language Models
-------------
The demo uses the `gpt-3.5-turbo-1106` model from OpenAI by default. This model doesn't produce the best
results, but it's about 10x faster than GPT-4, and also cheaper to use. You can change the model used
by setting the `fable_saga.default_openai_model_name` parameter to the model of your choice. You can also go with another model
provider supported by LangChain and pass that in when creating the agent.

The demo uses OpenAI because it's the easiest to get started with, is very effective, and supports json output well.
SAGA parses the output as JSON, so you can use any model that outputs JSON. If you use a different model, you may need
to add more specifics to the context that the model should only generate valid JSON or fine tune the model to
produce better results.

Installing as a PyPi Package
-------------
This will be coming once development settles a little first.


Installing Dependencies
==================

Python on Windows
----
Install Python 3.11.4
* Download and Install [Python 3.11.4](https://www.python.org/downloads/release/python-3114/) (earlier Python 3 versions may also work)
* Add python to your **Path** environment variable (e.g. `C:\Python\Python311\python.exe`)

Python on OSX and Linux
---
This is generally easier but platform specific. We recommend using homebrew on OSX and your integrated manager on linux (e.g. apt-get).


Install Poetry and Dependencies
-----
* Download and install the latest [Poetry](https://python-poetry.org/docs/#installing-with-the-official-installer)
* Add poetry to your **Path** environment variable (e.g. `C:\Users\Username\AppData\Roaming\pypoetry\venv\Scripts`)
* Run this command in the fable-generative-agents repo: `poetry install`

Configure OpenAI
---------
In order to avoid having to pass the OPENAI_API_KEY to the demo every time, you can set it as an environment variable.
* WINDOWS: Create a System environment variable called **OPENAI_API_KEY** and paste your API key into the value field
* ELSEWHERE: Creating a .env file or adding it to your bash profile should work. Many IDEs also have a way to set environment variables.

Install PyCharm (optional)
--------
Just a recommendation for our internal team who want to get this setup. Community folks can use any IDE you want.
* Download and install the latest [PyCharm Community Edition](https://www.jetbrains.com/pycharm/download/other.html)
* Open fable-generative-agents in PyCharm
* Configure interpreter: 
  - Settings > Project > Python Interpreter > Add Interpreter > Add Local Interpreter > Poetry Environment
  - Base interpreter: `C:\Users\<Username>\AppData\Roaming\pypoetry\venv\Scripts\python.exe`
  - Check Install packages from pyproject.toml
  - Poetry executable: `C:\Users\<Username>\AppData\Roaming\Python\Scripts\poetry.exe`
  - Choose 'OK'





CHANGELOG.md
v0.2.0
-------
- Added `fable_saga.Agent` class to make it easier to use SAGA in your own simulations.
- Simplified fable_saga itself to make it easier to use.
- Moved most of the demo code to an included open-source demo project: space_colony
- Added some documentation to get started with the demo and the basic concepts of SAGA.
- Removed the socketio code from the demo. It will be added back in soon.

v0.1.0
-------
- Initial release for use with Fable's Thistle Gulch demo.

fable_saga/server.py
import json
import logging
from typing import List, Optional

import cattrs
from attr import define
from aiohttp import web
import socketio
from cattrs import structure, unstructure
from langchain.chat_models.base import BaseLanguageModel

import fable_saga

logger = logging.getLogger(__name__)

"""
Sets up a server that can be used to generate actions for SAGA. Either HTTP or socketio can be used.

For HTTP, make a POST request to the server with a JSON body of the following format:
For socketio, send a message with the following format to the "generate-actions" event:
{
  "context": "You are a mouse",
  "skills": [{
    "name": "goto",
    "description": "go somewhere",
    "parameters": {
      "location": "<str: where you want to go>"
    }
  }]
}

Response:




"""



@define(slots=True)
class ActionsRequest:
    """Request to generate actions."""
    context: str
    skills: List[fable_saga.Skill] = {}
    retries: int = 0
    verbose: bool = False
    reference: Optional[str] = None


@define(slots=True)
class ActionsResponse:
    """Response from generating actions."""
    actions: Optional[fable_saga.GeneratedActions] = None
    error: str = None
    reference: Optional[str] = None


class SagaServer:
    """ Server for SAGA. """
    def __init__(self, llm: BaseLanguageModel = None):
        super().__init__()
        if llm is None:
            llm = fable_saga.ChatOpenAI()
        self.agent = fable_saga.Agent(llm)

    async def generate_actions(self, req: ActionsRequest) -> ActionsResponse:
        # Generate actions
        try:
            assert isinstance(req, ActionsRequest), f"Invalid request type: {type(req)}"
            actions = await self.agent.generate_actions(req.context, req.skills, req.retries, req.verbose)
            response = ActionsResponse(actions=actions, reference=req.reference)
            if actions.error is not None:
                response.error = f"Generation Error: {actions.error}"
            return response
        except Exception as e:
            logger.error(str(e))
            return ActionsResponse(actions=None, error=str(e), reference=req.reference)


if __name__ == '__main__':

    # Parse command line arguments
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--type', type=str, required=True, help="Type of server to run.", choices=["socketio", "http"])
    parser.add_argument('--host', type=str, default='localhost', help='Host to listen on')
    parser.add_argument('--port', type=int, default=8080, help='Port to listen on')
    parser.add_argument('--cors', type=str, default=None, help='CORS origin')
    args = parser.parse_args()

    # Create common server objects
    server = SagaServer()
    app = web.Application()

    # Create socketio server
    if args.type == 'socketio':
        if args.cors is None:
            args.cors = '*'
        sio = socketio.AsyncServer(async_mode='aiohttp', cors_allowed_origins=args.cors)
        sio.attach(app)

        @sio.event
        def connect(sid, environ):
            logger.info("connect:" + sid)

        @sio.event
        def disconnect(sid):
            logger.info("disconnect:" + sid)

        @sio.on('*')
        def catch_all(event, sid, *data):
            """Catch all unhandled events that have one message. (common)"""
            logger.error(f"Unhandled event: {event} {sid} {data}")

        @sio.on('generate-actions')
        async def generate_actions(sid, message_data: str):
            logger.debug(f"Request from {sid}: {message_data}")
            try:
                data = json.loads(message_data)
                request = structure(data, ActionsRequest)
                actions = await server.generate_actions(request)
                response = unstructure(actions)
                logger.debug(f"Response: {response}")
                return response
            except cattrs.errors.ClassValidationError as e:
                error = f"Error validating request: {json.dumps(cattrs.transform_error(e))}"
            except Exception as e:
                error = str(e)
            logger.error(error)
            return unstructure(ActionsResponse(actions=None, error=error))

    # Create HTTP server
    elif args.type == 'http':
        """HTTP server
        Make a POST request to the server with a JSON body of the following format:
        {
          "type" : "generate-actions",
          "data" : {
            "context": "You are a mouse",
            "skills": [{
              "skill": "goto",
              "parameters": {
                "location": "<str: where you want to go>"
              }
            }]
          }
        }
        
        Response:
        {
          "options": [
            {
              "skill": "goto",
              "parameters": {
                "location": "cheese pantry"
              }
            },
            ...
          ],
          "scores": [
            0.8,
            ...
          ]
        }
        """


        async def generate_actions(request):
            """Handle POST requests to the server."""
            logger.debug(f"Request : {request}")
            params = await request.json()
            try:
                actions_req: ActionsRequest = structure(params, ActionsRequest)
                actions_response = await server.generate_actions(actions_req)
                response = unstructure(actions_response)
                logger.debug(f"Response: {response}")
                return web.json_response(response)
            except cattrs.errors.ClassValidationError as e:
                error = f"Error validating request: {json.dumps(cattrs.transform_error(e))}"
            except Exception as e:
                error = str(e)
            logger.error(error)
            return web.json_response(unstructure(ActionsResponse(actions=None, error=error, reference=params.get('reference'))))


        # Listen for POST requests on the root path
        app.add_routes([web.post('/', generate_actions)])

    else:
        raise ValueError("Invalid server type: " + args.type)

    # Setup logging
    formatter = logging.Formatter('%(asctime)s - saga.server - %(levelname)s - %(message)s')
    handler = logging.StreamHandler()
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    logger.setLevel(logging.INFO)

    # Run server
    web.run_app(app, host=args.host, port=args.port)
